crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1359 @@
+                );
+            }
+        };
+
+        trace!("     | *filestat_ptr={:?}", host_filestat);
+
+        Ok(host_filestat)
+    }
+
+    fn fd_filestat_set_size(&self, fd: types::Fd, size: types::Filesize) -> Result<()> {
+        trace!("fd_filestat_set_size(fd={:?}, size={})", fd, size);
+
+        let entry = unsafe { self.get_entry(fd)? };
+        let file = entry
+            .as_descriptor(types::Rights::FD_FILESTAT_SET_SIZE, types::Rights::empty())?
+            .as_file()?;
+        // This check will be unnecessary when rust-lang/rust#63326 is fixed
+        if size > i64::max_value() as u64 {
+            return Err(Errno::TooBig);
+        }
+        match file {
+            Descriptor::OsHandle(fd) => fd.set_len(size)?,
+            Descriptor::VirtualFile(virt) => virt.filestat_set_size(size)?,
+            _ => {
+                unreachable!(
+                    "implementation error: fd should have been checked to not be a stream already"
+                );
+            }
+        };
+        Ok(())
+    }
+
+    fn fd_filestat_set_times(
+        &self,
+        fd: types::Fd,
+        atim: types::Timestamp,
+        mtim: types::Timestamp,
+        fst_flags: types::Fstflags,
+    ) -> Result<()> {
+        trace!(
+            "fd_filestat_set_times(fd={:?}, atim={}, mtim={}, fst_flags={})",
+            fd,
+            atim,
+            mtim,
+            fst_flags
+        );
+        let entry = unsafe { self.get_entry(fd)? };
+        let fd = entry
+            .as_descriptor(types::Rights::FD_FILESTAT_SET_TIMES, types::Rights::empty())?
+            .as_file()?;
+        fd::filestat_set_times_impl(&fd, atim, mtim, fst_flags)
+    }
+
+    fn fd_pread(
+        &self,
+        fd: types::Fd,
+        iovs: &types::IovecArray<'_>,
+        offset: types::Filesize,
+    ) -> Result<types::Size> {
+        trace!("fd_pread(fd={:?}, iovs={:?}, offset={})", fd, iovs, offset,);
+
+        let mut buf = Vec::new();
+        let mut bc = GuestBorrows::new();
+        bc.borrow_slice(iovs)?;
+        for iov_ptr in iovs.iter() {
+            let iov_ptr = iov_ptr?;
+            let iov: types::Iovec = iov_ptr.read()?;
+            let slice = unsafe {
+                let buf = iov.buf.as_array(iov.buf_len);
+                let raw = buf.as_raw(&mut bc)?;
+                &mut *raw
+            };
+            buf.push(io::IoSliceMut::new(slice));
+        }
+
+        let mut entry = unsafe { self.get_entry_mut(fd)? };
+        let file = entry
+            .as_descriptor_mut(
+                types::Rights::FD_READ | types::Rights::FD_SEEK,
+                types::Rights::empty(),
+            )?
+            .as_file_mut()?;
+
+        if offset > i64::max_value() as u64 {
+            return Err(Errno::Io);
+        }
+
+        let host_nread = match file {
+            Descriptor::OsHandle(fd) => {
+                let pos = SeekFrom::Start(offset);
+                fd.seek(pos)?;
```

Comment:
In terms of semantics, I'm not sure if `pread` is specified to change the file cursor or not, but I believe this implementation does? Might be good to double check...

Reply:
Oh crap, you're right, it shouldn't! I forgot to rewind the cursor! Thanks for bringing it up!

Reply:
Fixed! Thanks for spotting that one!